### ISSUE TITLE
22.3-7 Stack-based DFS time complexity

### ISSUE THREAD

**Issue by i-to** [walkccc/CLRS#329]
Link to the solution: https://walkccc.github.io/CLRS/Chap22/22.3/#223-7

Current stack-based solution is not exactly equivalent to the recursive one. 

When recursive dfs returns from the recursive call, it proceeds immediately to the next vertex in the parent's neighbor list. On the other hand, stack-based version will scan the neighbor list from the beginning each time to find the first white vertex.

I think, this is not guaranteed to be have linear time complexity for _any_ input. As a counter example, consider searching star-like graph from center.

**Comment by walkccc** [comment]
Yes, I agree with you. The problem asks us to eliminate recursion. To achieve the same time complexity, we have to use a variable to record who's the first WHITE vertex `v` of a parent `u` (suppose we are going to explore `(u, v)`), so that next time we can proceed with the next vertex in the `u`'s neighbor list.

I updated the solution and the C++ code a little bit for better readability.

Thanks.

**Comment by aetilley** [comment]
I'd also be interested in the best way to do this.   

You could use a pointer like you said to the next white element, but this pointer must be updated automatically each time the pointed-to element is set to grey.

It might be best to have FIRST-WHITE-NEIGHBOR(G, u) consult a map taking each u to a collection of its white neighbors.  But this collection must

1) Be able to pop some (any) member in O(1) and,
2) Support an efficient Remove(v) operation to be called when member v is set to grey (Remove would be called once for each edge in G.E).
